pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -360,33 +361,72 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 			//
 			// See https://github.com/kubernetes/kubernetes/pull/42147#discussion_r103357248
 		}
-		for i := 0; i < cidrUpdateRetries; i++ {
-			if err = utilnode.PatchNodeCIDRs(ca.client, types.NodeName(node.Name), cidrStrings); err == nil {
-				klog.InfoS("Set the node PodCIDRs", "nodeName", node.Name, "cidrStrings", cidrStrings)
-				break
-			}
-		}
+		node.Spec.PodCIDR = cidrStrings[0]
+		node.Spec.PodCIDRs = cidrStrings
 	}
+
+	err = ca.updateNodeCIDR(node, oldNode)
 	if err != nil {
-		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRAssignmentFailed")
-		klog.ErrorS(err, "Failed to update the node PodCIDR after multiple attempts", "nodeName", node.Name, "cidrStrings", cidrStrings)
 		return err
 	}
-
-	if northInterfaces != nil || additionalNodeNetworks != nil {
-		if err := ca.updateMultiNetworkAnnotations(node, northInterfaces, additionalNodeNetworks); err != nil {
+	if !reflect.DeepEqual(node.Annotations, oldNode.Annotations) {
+		if err = utilnode.PatchNodeMultiNetwork(ca.client, node); err != nil {
+			nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRAssignmentFailed")
+			klog.ErrorS(err, "Failed to update the node annotations and capacity for multi-networking", "nodeName", node.Name)
 			return err
 		}
 	}
-	err = utilnode.SetNodeCondition(ca.client, types.NodeName(node.Name), v1.NodeCondition{
-		Type:               v1.NodeNetworkUnavailable,
-		Status:             v1.ConditionFalse,
-		Reason:             "RouteCreated",
-		Message:            "NodeController create implicit route",
-		LastTransitionTime: metav1.Now(),
+	return err
+}
+
+func (ca *cloudCIDRAllocator) setNetworkCondition(node *v1.Node) {
+	for i := range node.Status.Conditions {
+		if node.Status.Conditions[i].Type == v1.NodeNetworkUnavailable {
+			// we do not update Times so that we do not trigger unnecessary updates
+			node.Status.Conditions[i].Status = v1.ConditionFalse
+			node.Status.Conditions[i].Reason = "RouteCreated"
+			node.Status.Conditions[i].Message = "NodeController create implicit route"
+			return
+		}
+	}
+	// NodeNetworkUnavailable condition not found, lets add it to trigger update,
+	// proper values are set during Patch.
+	node.Status.Conditions = append(node.Status.Conditions, v1.NodeCondition{
+		Type: v1.NodeNetworkUnavailable,
 	})
-	if err != nil {
-		klog.ErrorS(err, "Error setting route status for the node", "nodeName", node.Name)
+}
+
+func (ca *cloudCIDRAllocator) updateNodeCIDR(node, oldNode *v1.Node) error {
+	var err error
+
+	// update Spec.podCIDR
+	if !reflect.DeepEqual(node.Spec, oldNode.Spec) {
```

Comment:
ok, this is not correct, because you patch only the `node.Spec.PodCIDRs`, it may happen that you change the annotations or anything else and not the PodCIDRs, and because this is different you'll try to patch teh node cidrs when this is needed.

Reply:
that is why I am comparing `node.Spec`, not whole `node`, what is incorrect here ?

Reply:
oh, sorry, I didn't realize you only change the PodCIDRs on the spec